crates/wasi-experimental-http/src/lib.rs

```diff
@@ -22,34 +22,43 @@ pub fn request(req: Request<Option<Bytes>>) -> Result<Response<Bytes>, Error> {
     Ok(res.body(Bytes::from(body))?)
 }
 
-// TODO
-// This is a very inneficient way of getting a String from a `HeaderMap`.
 pub fn header_map_to_string(hm: &HeaderMap) -> Result<String, Error> {
-    let mut res: HashMap<String, String> = HashMap::new();
-    for (k, v) in hm.iter() {
-        res.insert(k.to_string(), v.to_str()?.to_string());
+    let mut res = String::new();
+    for (name, value) in hm
+        .iter()
+        .map(|(name, value)| (name.as_str(), std::str::from_utf8(value.as_bytes())))
+    {
+        let value = value?;
+        anyhow::ensure!(
+            !name
+                .chars()
+                .any(|x| x.is_control() || "(),/:;<=>?@[\\]{}".contains(x)),
+            "Invalid header name"
+        );
+        anyhow::ensure!(
+            !value.chars().any(|x| x.is_control()),
+            "Invalid header value"
+        );
+        res.push_str(&format!("{}:{}\n", name, value));
     }
-    let res = serde_json::to_value(&res)?.to_string();
     Ok(res)
 }
 
-// TODO
-// This is a very inneficient way of getting a `HeaderMap` from a `String.
-pub fn string_to_header_map(hm: String) -> Result<HeaderMap, Error> {
-    let hm: HashMap<String, String> = serde_json::from_str(&hm)?;
+pub fn string_to_header_map(s: &str) -> Result<HeaderMap, Error> {
     let mut headers = HeaderMap::new();
-    for (k, v) in hm.iter() {
+    for entry in s.lines() {
+        let (k, v) = entry.split_once(':').ok_or(anyhow::format_err!(
```

Comment:
This feature seems to be still in the nightly toolchain version (I think it might land in 1.52?):

```
error[E0658]: use of unstable library feature 'str_split_once': newly added
  --> crates/wasi-experimental-http/src/lib.rs:50:28
   |
50 |         let (k, v) = entry.split_once(':').ok_or(anyhow::format_err!(
   |                            ^^^^^^^^^^
   |
   = note: see issue #74773 <rust-lang/rust/issues/74773> for more information
```

Reply:
Oops, you're right, it hasn't been stabilized yet.

Fixed, thanks!